### DISCUSSION TITLE
`x-forwarded-host` header with value  does not match `origin` header.

### DISCUSSION THREAD

**Discussion by stefanosandes** [vercel/next.js#62050] [Help]
### Summary

Hello, everyone!

I have a Next application behind a proxy using Vercel's guide to create [multi zones](https://vercel.com/guides/how-can-i-serve-multiple-projects-under-a-single-domain).

In the root project, I have a Next application with rewrites configured in `next.config.js`. In the other project, I have another Next application with the `basePath` set to a subpath on the main domain.

Everything works great. Assets are loaded correctly, and images work with a small workaround.

The problem arises with Server Actions. When I try to call a server action, I receive the following error:

```
x-forwarded-host header with value <my_vercel_project_domain> does not match origin 
header with value <my_custom_domain> from a forwarded Server Actions request.
```

The general idea is to have multiple applications working on subpaths of a main domain.

The recommendation in these cases is to add my custom domain to the `allowedOrigins` configuration. However, the problem is that my application will have multiple domains assigned to it, one domain per customer, and I can't hardcode a new domain for every new customer.

I tried using a wildcard there (and I could handle security measures in middleware where I have a list of domains from the database), but as shown in this code snippet, it is not allowed: https://github.com/vercel/next.js/blob/0c21654845f78717a8198a69d968671477adc7f8/packages/next/src/server/app-render/csrf-protection.ts#L25-L39

Has anyone encountered or solved this kind of scenario?

**Comment by tselmek** [comment]
I have this exact same issue with Next 14.2.6. I've tried adding `<my_vercel_project_domain>` and `<my_custom_domain>` to the list of allowed forwarded hosts and allowed origins in my Next config but the error is still here.

**Comment by tapan-storetools** [comment]
I've been struggling with the same issue for the past few days. Today i finally found a solution which includes creating a middleware in nextjs

```typescript
import { NextApiResponse } from "next";
import { NextResponse, NextRequest } from "next/server";


type NextApiHandler = (req: NextRequest, res: NextApiResponse) => Promise<void>;

const actionHeaderCheckOverride = async (
  req: NextRequest,
  res: NextApiResponse,
  next: NextApiHandler,
): Promise<any> => {
  console.debug("REQUEST HEADERS:::: ", req.headers);

  
  const response = NextResponse.next();
  response.headers.set(
    "x-forwarded-host",
    req.headers.get("origin")?.replace(/(http|https):\/\//, "") || "*",
  );
  return response;
}
```

The above code rewrites the header `x-forwarded-host` with req.origin , which essentially matches origin with x-host-forwarded header. So that your nextjs action can execute after passing the check.
Hope this helps :)

**Reply by jelius-sama** [reply]
I was dealing with the same issue when I worked on my project on GitHub Codespace. I reviewed the documentation and implemented it in my code, but it did not work as expected. After conducting some research (which provided limited insights) and carefully analyzing the error message:
`
x-forwarded-host header with value <github_codespace_domain> does not match origin header with value localhost:<port> from a forwarded Server Actions request.
`

I added a conditional check in my `middleware.ts` file to set the response header to `localhost:<port>` when the environment is set to development. This resolved the issue. However, I am uncertain whether this approach is secure for a production environment. Below is the code snippet:
```ts
if (process.env.NODE_ENV !== "production") {
    response.headers.set('x-forwarded-host', 'localhost:5500');
}
```

**Reply by kingsonseang** [reply]
> I was dealing with the same issue when I worked on my project on GitHub Codespace. I reviewed the documentation and implemented it in my code, but it did not work as expected. After conducting some research (which provided limited insights) and carefully analyzing the error message: `x-forwarded-host header with value <github_codespace_domain> does not match origin header with value localhost:<port> from a forwarded Server Actions request.`
> 
> I added a conditional check in my `middleware.ts` file to set the response header to `localhost:<port>` when the environment is set to development. This resolved the issue. However, I am uncertain whether this approach is secure for a production environment. Below is the code snippet:
> 
> ```ts
> if (process.env.NODE_ENV !== "production") {
>     response.headers.set('x-forwarded-host', 'localhost:5500');
> }
> ```

Thanks a lot, had the same issue with Codespaces, Not sure if this helps but here's my implementation based on @jelius-sama comment.

PS: I'm using Auth.js for authentication, you can just follow the same pattern:

```typescript
import { auth } from '@/auth';
import { NextResponse } from 'next/server';

export default auth((request) => {
  if (process.env.NODE_ENV !== 'production') {
    // Create a new response and set the header on it
    const response = NextResponse.next();
    response.headers.set('x-forwarded-host', 'localhost:4000');
    return response;
  }

  // Return NextResponse.next() to continue
  return NextResponse.next();
});

export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|logo|media|icons|lottie|static|favicon.ico|swenv.js|firebase-messaging-sw.js|service-worker.js|swenv.js|sitemap.xml|robots.txt).*)',
  ],
};
```

The key fix is using `NextResponse.next()` and setting headers on the response instead of trying to modify the request directly, which isn't allowed in Next.js 15 middleware.

**Comment by dytra** [comment]
I'm using codespaces and fixed by creating middleware.ts file 

```typescript
// middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(req: NextRequest) {
  const url = req.nextUrl;

  if (req.method === 'POST' && url.pathname === '/') {
    const headers = new Headers(req.headers);
    const host = 'super-duper-tribble-wg9r6r749w2xxw-3000.app.github.dev';
    headers.set('origin', `https://${host}`);
    headers.set('x-forwarded-host', host);
    return NextResponse.next({ request: { headers } });
  }

  return NextResponse.next();
}

```
